lib/dolly/mango.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module Dolly
+  module Mango
+
+    SELECTOR_SYMBOL = "$"
+    AVAILABLE_SELECTORS = %I[
+      lt
+      lte
+      eq
+      ne
+      gte
+      gt
+      exists
+      type
+      in
+      nin
+      size
+      mod
+      regex
+    ].freeze
+
+    DESIGN = "_find"
+
+    def find_by(query, opts = {})
+      opts.merge!(limit: 1)
+      perform_query(build_query(query, opts))[:docs].first
+    end
+
+    def where(query, opts = {})
+      perform_query(build_query(query, opts))[:docs]
+    end
+
+    private
+
+    def perform_query(structured_query)
+      connection.post(DESIGN, structured_query)
+    end
+
+    def build_query(query, opts)
+      {
+        "selector" => build_selectors(query)
+      }.merge(opts)
+    end
+
+    def build_selectors(query)
+      query.map do |key, values|
+        inner_key = values.keys.first
+        inner_value = values.values.first
+        next unless AVAILABLE_SELECTORS.include?(inner_key)
+
+        {
+          "#{key}" => {
+            build_key(inner_key) => inner_value
+          }
+        }
+      end.compact.inject(:merge)
```

Comment:
`map` are usually faster than each_with_object, but since there are 3 iterations here, maybe there is some gaing on doing everything in once?

Reply:
@javierg You are right, with the following implementation these were the results
```
def construct_selectors(query)
      query.each_with_object(Hash.new) do |(key, value), hsh|
        inner_key = value.keys.first
        inner_value = value.values.first
        next unless AVAILABLE_SELECTORS.include?(inner_key)
        hsh.merge!({
          "#{key}" => {
            build_key(inner_key) => inner_value
          }
        })
      end
    end
```
`=> [#<Benchmark::Tms:0x00007fd1375e84a8 @label="", @real=2.9999995604157448e-05, @cstime=0.0, @cutime=0.0, @stime=3.0000000000030003e-06, @utime=3.200000000000425e-05, @total=3.500000000000725e-05>]`

The previous implementation had these results:

`=> [#<Benchmark::Tms:0x00007fd137338d20 @label="", @real=5.300000339047983e-05, @cstime=0.0, @cutime=0.0, @stime=1.4000000000000123e-05, @utime=4.500000000000337e-05, @total=5.9000000000003494e-05>]`